Do driver registration through a function.

danielballan opened the pull request:
Suggested by @HarinarayanKrishnan

Goals:
* Make the answer to "How do I register a driver at run time?" easier to discover.
* Ensure that two different code bases do not silently step on each other. This can cause trouble:
  ```py
  intake.registry['thing'] = A
  # ... later, perhaps in a different package...
  intake.registry['thing'] = B  # silently changed the registry under us!


This PR attempts to address both of these by introducing a new function `intake.register_driver` and laying a path toward making `intake.registry` an immutable view in the future. It works like this, a nice pattern that I have employed a couple times before to make shared registries safer and avoid confusing bugs.

```py
register_driver('thing', A)
register_driver('thing', A)  # OK, this is redundant but harmless
register_driver('thing', B)  # raises ValueError
register_driver('thing', B, overwrite= True)  # OK, since you have acknowledged that you are stepping on something
```

The code has been updated to use `register_driver` internally, never modifying `intake.registry` directly. Track is laid to make `intake.registry` an immutable view in the future; however, other libraries are likely modifying it directly now, so to start mutation is _support_ but issues a `DeprecationWarning`.

Unfortunately, as noted in code comments, we have to use `overwrite=True` internally in some situations. See #477. But this is should help avoid libraries accidentally stepping on each other and many cases.

danielballan commented on intake/source/__init__.py:
```diff
@@ -8,8 +8,60 @@
+    _registry[name] = driver
+
+
+def unregister_driver(name):
```
I am not sure whether we actually need or want this one. Letting a package silently _remove_ a driver from underneath us isn't as likely to cause confusing bugs as silently _change_ a driver from underneath us, but it's still not great and I'm not sure there is a use case for this. Maybe `register_driver` with `overwrite` is enough.

martindurant replied:
It is useful, at least for debugging. I don't expect too many end-users to need it, but I don't think it hurts.

danielballan replied:
Ok, that works for me.

martindurant commented:
On (re)importing source, we could do something a little shady like 
```
exec("import {modname}; cls = {modname}.{clsname}")
```
I'm not sure how you protect from code injection here (require only [a-zA-Z0-9.]?), but it doesn't incur multiple import costs, since it will use the module cache.

danielballan commented:
I believe that in all situations where `exec("import {modname}")` would work, `importlib.import_module(modname)` would also work. The thing is, intake currently supports exec-ing `.py` files at arbitrary locations outside of the `sys.path`, where neither of the approaches mentioned here would work.

danielballan commented:
The easier thing to do here would be to stop supporting that. Are you aware of an important use case?

martindurant commented:
Probably no reason to support it, agreed; other tricks like modifying path are not appetising.

On March 9, 2020 3:29:29 PM EDT, Dan Allan <notifications@github.com> wrote:
>The easier thing to do here would be to stop supporting that. Are you
>aware of an important use case?
>
>-- 
>You are receiving this because your review was requested.
>Reply to this email directly or view it on GitHub:
>https://github.com/intake/intake/pull/478#issuecomment-596736721

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

martindurant commented on intake/catalog/local.py:
```diff
@@ -512,7 +512,11 @@ def register_plugin_module(mod):
+            # load_plugins_from_module can employ imp.load_source in which case
+            # we will re-exec the code and get a different type each time, so
+            # we have to tolerate overwriting here.
+            uses_load_source = mod.endswith('.py')
```
Yep, suggest raising an error for this case and dropping in `register_driver`. In hindsight, it's quite a security hole to automatically execute code just because you opened a catalogue, right? Also, it won't work for remote catalogues.

danielballan replied:
Both good points. Will do.

danielballan replied:
Done in 6e9b2b1c2110b3bfa55a95bb5496b43d82b29e84

martindurant commented:
Any thoughts on `intake.container.container_map`?

danielballan commented:
Sure, might as well address `container_map` in the same PR. On it....

danielballan commented:
Some panel-related errors have appeared that seem unrelated to this PR. Was there a release recently?

danielballan commented:
Reproduced panel test-related issues locally and fixed. Summary:

- [x] Made `intake.registry` warn if directly mutated, with a path to making it immutable in a couple releases. Added `intake.register_driver` and `intake.unregister_driver`, 
- [x] Same for `intake.container.container_map`, `intake.container.register_container`, `intake.container.unregister_container`.
- [x] Removed ability to execute arbitrary Python scripts when a catalog is opened. If someone tries to use this feature, a specific error is raised explaining that this has intentionally changed and what to do installed. Closes #477; we no longer user the deprecated `imp` module.

martindurant commented:
intake-xarray is the only place where we need register_container?

I'm very happy to see this go in. Any last comments @jsignell or anyone else?

jsignell commented:
This all looks very sane and like a nice cleanup.
